File: src/0205-Isomorphic-Strings/isomorphic_strings.py
from typing import Dict

class Solution:
    def isIsomorphic(self, s: str, t: str) -> bool:
        if len(s) != len(t):
            return False
        
        s_to_t: Dict[str, str] = {}
        t_to_s: Dict[str, str] = {}
        
        for char_s, char_t in zip(s, t):
            if char_s in s_to_t:
                if s_to_t[char_s] != char_t:
                    return False
            else:
                s_to_t[char_s] = char_t
            
            if char_t in t_to_s:
                if t_to_s[char_t] != char_s:
                    return False
            else:
                t_to_s[char_t] = char_s
        
        return True
    
if __name__ == "__main__":
    s: str = "egg"
    t: str = "add"
    solution = Solution()
    
    print(solution.isIsomorphic(s, t))

File: src/0205-Isomorphic-Strings/isomorphic_strings.cpp
#include <iostream>
#include <unordered_map>
#include <string>

class Solution {
public:
    bool isIsomorphic(std::string s, std::string t) {
        if (s.length() != t.length()) {
            return false;
        }
        
        std::unordered_map<char, char> s_to_t;
        std::unordered_map<char, char> t_to_s;
        
        for (int i = 0; i < s.length(); ++i) {
            char char_s = s[i];
            char char_t = t[i];
            
            if (s_to_t.find(char_s) != s_to_t.end()) {
                if (s_to_t[char_s] != char_t) {
                    return false;
                }
            } else {
                s_to_t[char_s] = char_t;
            }
            
            if (t_to_s.find(char_t) != t_to_s.end()) {
                if (t_to_s[char_t] != char_s) {
                    return false;
                }
            } else {
                t_to_s[char_t] = char_s;
            }
        }
        
        return true;
    }
};

int main() {
    std::string s = "egg";
    std::string t = "add";
    Solution solution;
    
    std::cout << std::boolalpha << solution.isIsomorphic(s, t) << std::endl;
    
    return 0;
}